src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,81 +5,64 @@
+    if (mutableSamplerCount > 0)
+        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
 }
 
-core::smart_refctd_ptr<asset::IDescriptor>* IGPUDescriptorSet::getDescriptorStorage(const asset::E_DESCRIPTOR_TYPE type) const
+IGPUDescriptorSet::~IGPUDescriptorSet()
 {
-    core::smart_refctd_ptr<asset::IDescriptor>* baseAddress;
-    switch (type)
+    if (!isZombie())
     {
-    case asset::EDT_COMBINED_IMAGE_SAMPLER:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_textureStorage.get());
-        break;
-    case asset::EDT_STORAGE_IMAGE:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_storageImageStorage.get());
-        break;
-    case asset::EDT_UNIFORM_TEXEL_BUFFER:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UTB_STBStorage.get());
-        break;
-    case asset::EDT_STORAGE_TEXEL_BUFFER:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UTB_STBStorage.get()) + m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER];
-        break;
-    case asset::EDT_UNIFORM_BUFFER:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get());
-        break;
-    case asset::EDT_STORAGE_BUFFER:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER];
-        break;
-    case asset::EDT_UNIFORM_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER]);
-        break;
-    case asset::EDT_STORAGE_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC]);
-        break;
-    case asset::EDT_INPUT_ATTACHMENT:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_storageImageStorage.get()) + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE];
-        break;
-    case asset::EDT_ACCELERATION_STRUCTURE:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_accelerationStructureStorage.get());
-        break;
-    default:
-        assert(!"Invalid code path.");
-        return nullptr;
-    }
+        const bool allowsFreeing = m_pool->m_creationParameters.flags.hasFlags(IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT);
```

Comment:
why abuse the friendship for this, don't we have a `getFlags()` or `getCreationParams()` ?

Reply:
I made a `public` method by the same name in `IDescriptorPool` which checks the presence of that flag. Resoved?

Reply:
yep